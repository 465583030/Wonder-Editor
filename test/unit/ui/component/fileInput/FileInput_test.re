open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe(
    "FileInput ui component",
    (_) => {
      let sandbox = getSandboxDefaultVal();
      let _clickShowInputEvent = (domChildren) => {
        let btn = WonderCommonlib.ArraySystem.unsafeGet(domChildren, 0);
        btn##props##onClick()
      };
      let _changeTextAreaEvent = (value, domChildren) => {
        let div = WonderCommonlib.ArraySystem.unsafeGet(domChildren, 1);
        let textarea = WonderCommonlib.ArraySystem.unsafeGet(div##children, 0);
        EventToolUI.triggerChangeEvent(textarea, value)
      };
      let _submitClickEvent = (domChildren) => {
        let div = WonderCommonlib.ArraySystem.unsafeGet(domChildren, 1);
        let submitBtn = WonderCommonlib.ArraySystem.unsafeGet(div##children, 1);
        submitBtn##props##onClick()
      };
      beforeEach(() => sandbox := createSandbox());
      afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
      describe(
        "test workflow",
        () => {
          test(
            "set showInput button text",
            (_) => {
              let component = ReactTestRenderer.create(<FileInput buttonText="showInput" />);
              let json = ReactTestRenderer.toJSON(component);
              toMatchSnapshot(expect(json))
            }
          );
          test(
            "click the showInput button, show the textarea and submit-button",
            (_) => {
              let component = ReactTestRenderer.create(<FileInput buttonText="showInput" />);
              EventToolUI.triggerComponentEvent(component, _clickShowInputEvent);
              let json = ReactTestRenderer.toJSON(component);
              toMatchSnapshot(expect(json))
            }
          );
          test(
            "key in text",
            (_) => {
              let component = ReactTestRenderer.create(<FileInput buttonText="showInput" />);
              EventToolUI.triggerComponentEvent(component, _clickShowInputEvent);
              EventToolUI.triggerComponentEvent(
                component,
                _changeTextAreaEvent("you can input value in textarea")
              );
              let json = ReactTestRenderer.toJSON(component);
              toMatchSnapshot(expect(json))
            }
          );
          test(
            "click submit-button,the onSubmit method should be called",
            (_) => {
              let inputValue = "you can click submit after input value";
              let onSubmit = createEmptyStubWithJsObjSandbox(sandbox);
              let component =
                ReactTestRenderer.create(<FileInput buttonText="showInput" onSubmit />);
              EventToolUI.triggerComponentEvent(component, _clickShowInputEvent);
              EventToolUI.triggerComponentEvent(component, _changeTextAreaEvent(inputValue));
              EventToolUI.triggerComponentEvent(component, _submitClickEvent);
              onSubmit |> expect |> toCalledWith([inputValue])
            }
          )
        }
      );
      describe(
        "deal with the specific case",
        () => {
          test(
            "if submit method not pass in, shouldn't execute onSubmit method",
            (_) => {
              let onSubmit = createEmptyStubWithJsObjSandbox(sandbox);
              let component = ReactTestRenderer.create(<FileInput buttonText="showInput" />);
              EventToolUI.triggerComponentEvent(component, _clickShowInputEvent);
              EventToolUI.triggerComponentEvent(component, _changeTextAreaEvent("this is value"));
              EventToolUI.triggerComponentEvent(component, _submitClickEvent);
              onSubmit |> expect |> not_ |> toCalled
            }
          );
          test(
            "if input value == '', click submit button shouldn't execute onSubmit method",
            (_) => {
              let onSubmit = createEmptyStubWithJsObjSandbox(sandbox);
              let component =
                ReactTestRenderer.create(<FileInput buttonText="showInput" onSubmit />);
              EventToolUI.triggerComponentEvent(component, _clickShowInputEvent);
              EventToolUI.triggerComponentEvent(component, _changeTextAreaEvent(""));
              EventToolUI.triggerComponentEvent(component, _submitClickEvent);
              onSubmit |> expect |> not_ |> toCalled
            }
          )
        }
      )
    }
  );