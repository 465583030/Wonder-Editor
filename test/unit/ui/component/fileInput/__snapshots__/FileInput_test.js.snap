// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput ui component test workflow click the showInput button, show the textarea and submit-button 1`] = `
<div
  className="file-input"
>
  <button
    onClick={[Function]}
  >
    showInput
  </button>
  <div>
    <textarea
      className="ant-input number-input-input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      submit
    </button>
  </div>
</div>
`;

exports[`FileInput ui component test workflow key in text 1`] = `
<div
  className="file-input"
>
  <button
    onClick={[Function]}
  >
    showInput
  </button>
  <div>
    <textarea
      className="ant-input number-input-input"
      onChange={[Function]}
      type="text"
      value="you can input value in textarea"
    />
    <button
      onClick={[Function]}
    >
      submit
    </button>
  </div>
</div>
`;

exports[`FileInput ui component test workflow set showInput button text 1`] = `
<div
  className="file-input"
>
  <button
    onClick={[Function]}
  >
    showInput
  </button>
</div>
`;

exports[`fileInput ui component click the showInput button, show the textarea and submit button 1`] = `
<div
  className="file-input"
>
  <button
    onClick={[Function]}
  >
    showInput
  </button>
  <div>
    <textarea
      className="ant-input number-input-input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      submit
    </button>
  </div>
</div>
`;

exports[`fileInput ui component input text after click the showInput button 1`] = `
<div
  className="file-input"
>
  <button
    onClick={[Function]}
  >
    showInput
  </button>
  <div>
    <textarea
      className="ant-input number-input-input"
      onChange={[Function]}
      type="text"
      value="you can input value in textarea"
    />
    <button
      onClick={[Function]}
    >
      submit
    </button>
  </div>
</div>
`;

exports[`fileInput ui component set showInput button text 1`] = `
<div
  className="file-input"
>
  <button
    onClick={[Function]}
  >
    showInput
  </button>
</div>
`;
